<!DOCTYPE html>
<html lang="en">
<head>
    <meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
    <meta charset="utf-8">
    <!--<meta http-equiv="X-UA-Compatible" content="IE=edge">-->
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <meta name="description" content="">
    <meta name="author" content="">
    <link rel="shortcut icon" href="/img/favicon.png">
    <link rel="alternate" type="application/rss+xml" title="RSS" href="http://www.nextflow.io/feed.xml">
    <title>Nextflow - Workflows &amp; publishing: best practice for reproducibility </title>
    <meta name="keywords" content="dataflow, data-flow, dataflow programming model, flow-based programming, pipeline, computational pipeline, bioinformatics, bioinformatics pipeline, RNAseq, next gen sequencing, next generation sequencing, parallel computing, cloud computing, distributed computing, grid computing, cluster computing, Sun Grid Engine, SGE, Oracle Grid Engine"/>

    <!-- Bootstrap core CSS -->
    <link href="/css/bootstrap.css" rel="stylesheet">

    <!-- Custom styles for this template -->
    <link href="/css/color-styles.css" rel="stylesheet">
    <link href="/css/ui-elements.css" rel="stylesheet">
    <link href="/css/custom.css" rel="stylesheet">

    <!-- Resources -->
    <link href="/css/animate.css" rel="stylesheet">
    <link href="https://netdna.bootstrapcdn.com/font-awesome/4.0.3/css/font-awesome.css" rel="stylesheet">
    <link href='https://fonts.googleapis.com/css?family=Lobster' rel='stylesheet' type='text/css'>

    <!-- HTML5 shim and Respond.js IE8 support of HTML5 elements and media queries -->
    <!--[if lt IE 9]>
    <script src="https://oss.maxcdn.com/libs/html5shiv/3.7.0/html5shiv.js"></script>
    <script src="https://oss.maxcdn.com/libs/respond.js/1.3.0/respond.min.js"></script>
    <![endif]-->

</head>

<body class="body-green">

<div class="navbar navbar-inverse navbar-fixed-top" role="navigation">

    <div class="container">
        <div class="navbar-header">
            <a class="navbar-brand" href="/index.html" style='padding: 8px'><img src='/img/nextflow2014_no-bg.png' height='40px' style='border: 0;'></a>
        </div>
        <div class="navbar-collapse collapse">

            <ul class="nav navbar-nav navbar-left hidden-xs">
                <li class="show animated flipInX" >
                    <a href="/index.html#Features" class="pull-right">Features</a>
                </li>
                <li class="show animated flipInX" >
                    <a href="/index.html#GetStarted" class="pull-right">Quick start</a>
                </li>

                <li class="dropdown show animated flipInX">
                    <a href="#" class="dropdown-toggle" data-toggle="dropdown">Examples <b class="caret"></b></a>
                    <ul class="dropdown-menu">
                        <li><a href="/example1.html">Basic pipeline</a></li>
                        <li><a href="/example2.html">Mixing scripting languages</a></li>
                        <li><a href="/example3.html">BLAST pipeline</a></li>
                        <li><a href="/example4.html">RNA-Seq pipeline</a></li>
                        <li><a href="https://github.com/nextflow-io/examples" target="_blank">More examples..</a></li>
                    </ul>
                </li>

                <li class="show animated flipInX">
                    <a href="/docs/latest/index.html" class="pull-right" target="_blank">Documentation</a>
                </li>

                <li class="show animated flipInX">
                    <a href="/blog.html">Blog</a>
                </li>

                <li class="show animated flipInX">
                    <a href="/about-us.html">About Us</a>
                </li>
                   
            </ul>
            <ul  class="list-inline pull-right">
                <li class="show pull-right" style='position: relative; top: 10px; right: 75px' >
                	<a href='http://www.crg.eu' target='_blank' style='padding: 0'><img src='/img/crg_logo.png' height='45'/></a>
                </li>
            </ul>
        </div>
    </div>
</div>
	
    <!-- Body -->
	<div class="wrapper"> <!-- wrapper -->
	  <div class="container">
	    <div class="row">
		  <div class="col-sm-8">
			<div class="blg-summary">
			  <h3><a href="#">Workflows &amp; publishing: best practice for reproducibility </a></h3>
			  <ul class="text-muted list-inline blg-header">
				<li><i class="fa fa-user"></i> Evan Floden </li>
				<li><i class="fa fa-calendar"></i> 13 April 2016</li>
				<!--<li><i class="fa fa-comments-o"></i> 21 comments</li> -->
			  </ul>
			  <hr>
			  <div class="blg-text">
              <p>Publication time acts as a snapshot for scientific work. Whether a project is ongoing or not, work which was performed months ago must be described, new software documented, data collated and figures generated. </p><p>The monumental increase in data and pipeline complexity has led to this task being performed to many differing standards, or <a href="http://journals.plos.org/plosone/article?id=10.1371/journal.pone.0080278">lack of thereof</a>. We all agree it is not good enough to simply note down the software version number. But what practical measures can be taken?</p><p>The recent publication describing <em>Kallisto</em> <a href="http://dx.doi.org/10.1038/nbt.3519">(Bray et al. 2016)</a> provides an excellent high profile example of the growing efforts to ensure reproducible science in computational biology. The authors provide a GitHub <a href="https://github.com/pachterlab/kallisto_paper_analysis">repository</a> that <em>“contains all the analysis to reproduce the results in the kallisto paper”</em>. </p><p>They should be applauded and indeed - in the Twittersphere - they were. The corresponding author Lior Pachter stated that the publication could be reproduced starting from raw reads in the NCBI Sequence Read Archive through to the results, which marks a fantastic accomplishment.</p>
<blockquote class="twitter-tweet" data-lang="en"><p lang="en" dir="ltr">Hoping people will notice <a href="https://t.co/qiu3LFozMX">https://t.co/qiu3LFozMX</a> by <a href="https://twitter.com/yarbsalocin">@yarbsalocin</a> <a href="https://twitter.com/hjpimentel">@hjpimentel</a> <a href="https://twitter.com/pmelsted">@pmelsted</a> reproducing ALL the <a href="https://twitter.com/hashtag/kallisto?src=hash">#kallisto</a> paper from SRA→results</p>&mdash; Lior Pachter (@lpachter) <a href="https://twitter.com/lpachter/status/717279998424457216">April 5, 2016</a></blockquote>
<script async src="//platform.twitter.com/widgets.js" charset="utf-8"></script><p>They achieve this utilising the workflow framework <a href="https://bitbucket.org/snakemake/snakemake/wiki/Home">Snakemake</a>. Increasingly, we are seeing scientists applying workflow frameworks to their pipelines, which is great to see. There is a learning curve, but I have personally found the payoffs in productivity to be immense.</p><p>As both users and developers of Nextflow, we have long discussed best practice to ensure reproducibility of our work. As a community, we are at the beginning of that conversation - there are still many ideas to be aired and details ironed out - nevertheless we wished to provide a <em>state-of-play</em> as we see it and to describe what is possible with Nextflow in this regard.</p><h3>Guaranteed Reproducibility</h3><p>This is our goal. It is one thing for a pipeline to be able to be reproduced in your own hands, on your machine, yet is another for this to be guaranteed so that anyone anywhere can reproduce it. What I mean by guaranteed is that when a given pipeline is executed, there is only one result which can be output. Envisage what I term the <em>reproducibility triangle</em>: consisting of data, code and compute environment.</p><p><img src="/img/reproducibility-triangle.png" alt="Reproducibility Triangle""/></p><p><strong>Figure 1:</strong> The Reproducibility Triangle. <em>Data</em>: raw data such as sequencing reads, genomes and annotations but also metadata such as experimental design. <em>Code</em>: scripts, binaries and libraries/dependencies. <em>Environment</em>: operating system.</p><p>If there is any change to one of these then the reproducibililty is no longer guaranteed. For years there have been solutions to each of these individual components. But they have lived a somewhat discrete existence: data in databases such as the SRA and Ensembl, code on GitHub and compute environments in the form of virtual machines. We think that in the future science must embrace solutions that integrate each of these components natively and holistically.</p><h3>Implementation</h3><p>Nextflow provides a solution to reproduciblility through version control and sandboxing.</p><h4>Code</h4><p>Version control is provided via <a href="http://www.nextflow.io/docs/latest/sharing.html">native integration with GitHub</a> and other popular code management platforms such as Bitbucket and GitLab. Pipelines can be pulled, executed, developed, collaborated on and shared. For example, the command below will pull a specific version of a <a href="https://github.com/cbcrg/kallisto-nf">simple Kallisto + Sleuth pipeline</a> from GitHub and execute it. The <code>-r</code> parameter can be used to specify a specific tag, branch or revision that was previously defined in the Git repository. </p>
<pre><code>nextflow run cbcrg/kallisto-nf -r v0.9
</code></pre><h4>Environment</h4><p>Sandboxing during both development and execution is another key concept; version control alone does not ensure that all dependencies nor the compute environment are the same.</p><p>A simplified implementation of this places all binaries, dependencies and libraries within the project repository. In Nextflow, any binaries within the the <code>bin</code> directory of a repository are added to the path. Also, within the Nextflow <a href="https://github.com/cbcrg/kallisto-nf/blob/master/nextflow.config">config file</a>, environmental variables such as <code>PERL5LIB</code> can be defined so that they are automatically added during the task executions. </p><p>This can be taken a step further with containerisation such as <a href="http://www.nextflow.io/docs/latest/docker.html">Docker</a>. We have recently published <a href="https://doi.org/10.7717/peerj.1273">work</a> about this: briefly a <a href="https://github.com/cbcrg/kallisto-nf/blob/master/Dockerfile">dockerfile</a> containing the instructions on how to build the docker image resides inside a repository. This provides a specification for the operating system, software, libraries and dependencies to be run.</p><p>The images themself also have content-addressable identifiers in the form of <a href="https://docs.docker.com/engine/userguide/containers/dockerimages/#image-digests">digests</a>, which ensure not a single byte of information, from the operating system through to the libraries pulled from public repos, has been changed. This container digest can be specified in the <a href="https://github.com/cbcrg/kallisto-nf/blob/master/nextflow.config">pipeline config file</a>.</p>
<pre><code>process {
    container = &quot;cbcrg/kallisto-nf@sha256:9f84012739...&quot;
}
</code></pre><p>When doing so Nextflow automatically pulls the specified image from the Docker Hub and manages the execution of the pipeline tasks from within the container in a transparent manner, i.e. without having to adapt or modify your code. </p><h4>Data</h4><p>Data is currently one of the more challenging aspect to address. <em>Small data</em> can be easily version controlled within git-like repositories. For larger files the <a href="https://git-lfs.github.com/">Git Large File Storage</a>, for which Nextflow provides built-in support, may be one solution. Ultimately though, the real home of scientific data is in publicly available, programatically accessible databases. </p><p>Providing out-of-box solutions is difficult given the hugely varying nature of the data and meta-data within these databases. We are currently looking to incorporate the most highly used ones, such as the <a href="http://www.ncbi.nlm.nih.gov/sra">SRA</a> and <a href="http://www.ensembl.org/index.html">Ensembl</a>. In the long term we have an eye on initiatives, such as <a href="https://www.ncbi.nlm.nih.gov/bioproject/">NCBI BioProject</a>, with the idea there is a single identifier for both the data and metadata that can be referenced in a workflow.</p><p>Adhering to the practices above, one could imagine one line of code which would appear within a publication. </p>
<pre><code>nextflow run [user/repo] -r [version] --data[DB_reference:data_reference] -with-docker
</code></pre><p>The result would be guaranteed to be reproduced by whoever wished. </p><h3>Conclusion</h3><p>With this approach the reproducilbility triangle is complete. But it must be noted that this does not guard against conceptial or implemenation errors. It does not replace proper documentation. What it does is to provide transparency to a result.</p><p>The assumption that the deterministic nature of computation makes results insusceptible to irreproducbility is clearly false. We consider Nextflow with its other features such its polyglot nature, out-of-the-box portability and native support across HPC and Cloud environments to be an ideal solution in our everyday work. We hope to see more scientists adopt this approach to their workflows. </p><p>The recent efforts by the <em>Kallisto</em> authors highlight the appetite for increasing these standards and we encourage the community at large to move towards ensuring this becomes the normal state of affairs for publishing in science.</p><h3>References</h3><p>Bray, Nicolas L., Harold Pimentel, Páll Melsted, and Lior Pachter. 2016. “Near-Optimal Probabilistic RNA-Seq Quantification.” Nature Biotechnology, April. Nature Publishing Group. doi:10.1038/nbt.3519.</p><p>Di Tommaso P, Palumbo E, Chatzou M, Prieto P, Heuer ML, Notredame C. (2015) "The impact of Docker containers on the performance of genomic pipelines." PeerJ 3:e1273 doi.org:10.7717/peerj.1273.</p><p>Garijo D, Kinnings S, Xie L, Xie L, Zhang Y, Bourne PE, et al. (2013) "Quantifying Reproducibility in Computational Biology: The Case of the Tuberculosis Drugome." PLoS ONE 8(11): e80278. doi:10.1371/journal.pone.0080278</p>  
			  </div>
			  
			  <p class="tags">
				<a href="/tags/bioinformatics.html" class="background-color bg-hover-color">bioinformatics</a>
				<a href="/tags/reproducibility.html" class="background-color bg-hover-color">reproducibility</a>
				<a href="/tags/pipelines.html" class="background-color bg-hover-color">pipelines</a>
				<a href="/tags/nextflow.html" class="background-color bg-hover-color">nextflow</a>
				<a href="/tags/genomic.html" class="background-color bg-hover-color">genomic</a>
				<a href="/tags/docker.html" class="background-color bg-hover-color">docker</a>
			  </p>

			</div>
			<!-- COMMENTS -->
			<div class="comments">
			  <!--<h4>3 comments</h4>-->
			  
			    <div id="disqus_thread"></div>
                <script type="text/javascript">
                    if( window.location.hostname == 'localhost' || window.location.hostname == '127.0.0.1' ) { throw new Error('Skip Disqus on localhost') };
                    /* * * CONFIGURATION VARIABLES: EDIT BEFORE PASTING INTO YOUR WEBPAGE * * */
                    var disqus_shortname = 'nextflow'; // required: replace example with your forum shortname

                    /* * * DON'T EDIT BELOW THIS LINE * * */
                    (function() {
                        var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
                        dsq.src = '//' + disqus_shortname + '.disqus.com/embed.js';
                        (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
                    })();
                </script>
                <noscript>Please enable JavaScript to view the <a href="http://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
                <a href="http://disqus.com" class="dsq-brlink">comments powered by <span class="logo-disqus">Disqus</span></a>
    
			</div>
			 
			<!-- // Comments -->
		  </div>
		  <div class="col-sm-4">
			<h3>Stay Tuned <small>Social Links</small></h3>
			<hr>
			<ul class="blg-social">
			  <li>
			    <a href="/feed.xml">
			      <span class="icon rss">
				    <i class="fa fa-rss fa-2x"></i>
				  </span>
				  <span class="text-muted">Subscribe to our RSS Feed.</span>
				</a>
			  </li>
			  <li>
			    <a href="https://twitter.com/nextflowio">
			      <span class="icon twitter">
				    <i class="fa fa-twitter fa-2x"></i>
				  </span>
				  <span class="text-muted">Subscribe to our Twitter Feed.</span>
				</a>
			  </li>
			</ul>
            <!--
			<h3>Popular Stories</h3>
			<hr>
			<div class="block-inverse">
			  <ul>
				<li><a href="#">Lorem ipsum dolor sit amet, consectetur adipiscing elit.</a></li>
				<li><a href="#">Nullam id ipsum varius, tincidunt odio nec, placerat.</a></li>
				<li><a href="#">Sed sit amet auctor augue, nec dignissim ligula.</a></li>
				<li><a href="#">Lorem ipsum dolor sit amet, consectetur adipiscing elit.</a></li>
				<li><a href="#">Lorem ipsum dolor sit amet, consectetur adipiscing elit.</a></li>
			  </ul>
			</div>			
            -->
		  </div>
		</div>
	  </div>
	</div> <!-- / wrapper -->

	
<div class="footer-wrapper"> <!-- footer wrapper -->
    <hr>
    <div class="container">
        <footer>
            <ul class="list-inline pull-left">                         
                <li style="position: relative; top: 5px">
                  <a target='_blank' href="https://gitter.im/nextflow-io/nextflow?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge"><img src="https://badges.gitter.im/Join Chat.svg"/></a>
                </li>
            </ul>
            
            <ul class="list-inline pull-right">

                <li style="position: relative; top: 5px">
                    <a href="https://twitter.com/nextflowio" class="twitter-follow-button" data-show-count="false" >Follow @nextflowio</a>
                </li>

                <li>
                    <script type="text/javascript">
//<![CDATA[
<!--
var x="function f(x,y){var i,o=\"\",l=x.length;for(i=0;i<l;i++){if(i>(91+y))" +
"y*=2;y%=127;o+=String.fromCharCode(x.charCodeAt(i)^(y++));}return o;}f(\"=)" +
"3=+\\t\\016\\014C\\002M\\036N\\023\\037\\013\\031L\\004B\\000MSP_\\033\\031" +
"K\\017V\\025\\037\\025\\033\\t\\026,m?lh>qoaeo#t#mgqcQ|ppWc0u5*/4?\\\"\\025" +
"\\024\\013XPW_\\\\P\\002\\027S\\027A\\005\\022\\\\\\nOPUAU_\\020\\\\\\023@A" +
"@XP2i+~(hw|!ww{w$cby*=xi-x408(\\032(u7v[\\034\\020\\006\\020\\020\\024\\tH\\"+
"006D\\030\\031\\017\\035\\033\\002YB_\\033\\031_L\\005\\037RYUIL,^$?'\\rP$9" +
"c\\\" P\\027PlZR-\\014N\\\\^1<.b\\\"416RBL\\\"\\\\26>BbL22ZR.\\rN\\\\^3=-\\" +
"016\\\"\\\\74>BbJ32ZR.\\rN\\\\^HC,qNV57>BbK02ZR*\\rN\\\\^0?.b\\\"206RB{8?pZ" +
"R.\\017N\\\\^H&/\\017D/^$RB\\002H706RB\\007M:8n0:\\035Y,#o8,\\016\\\"\\\\v(" +
"DBb\\\"\\\\G2?.b\\\"J44>BbM12ZR.\\016N\\\\^OD^n4PT0?.b\\\"kr~\\177pG\\010|^" +
"$&xCEo\\\"h|kJ\\033r\\177='7\\025Uy*X'w\\026\\nAgba]L\\037ha(v6[\\032oAtov}" +
"\\023op` yP\\027rvU35QE7073;GEi?-w7\\nL=?o&xW\\005))-g%RBi963w\\026\\014od=" +
"fjY\\020en(v#RR\\\\ Z,#QRi\\\"tohEWy.~&x\\036\\020okrmpK\\030\\\"+\",91)"    ;
while(x=eval(x));
//-->
//]]>
</script>

                </li>

            </ul>
            <div class="clearfix"></div>
        </footer>
    </div>
</div>
<!-- / footer wrapper -->


<!-- Bootstrap core JavaScript
================================================== -->
<!-- Placed at the end of the document so the pages load faster -->
<script src="https://ajax.googleapis.com/ajax/libs/jquery/1.9.1/jquery.min.js"></script>
<script src="/js/bootstrap.min.js"></script>
<script src="/js/custom.js"></script>
<script>!function(d,s,id){var js,fjs=d.getElementsByTagName(s)[0],p=/^http:/.test(d.location)?'http':'https';if(!d.getElementById(id)){js=d.createElement(s);js.id=id;js.src=p+'://platform.twitter.com/widgets.js';fjs.parentNode.insertBefore(js,fjs);}}(document, 'script', 'twitter-wjs');</script>
<script>
  if( window.location.hostname == 'localhost' || window.location.hostname == '127.0.0.1' ) { throw new Error('Skip GA on localhost') };
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

  ga('create', 'UA-364526-10', 'auto');
  ga('send', 'pageview');
</script>

</body>
</html>